Lib/test/test_interpreters.py

```diff
@@ -0,0 +1,88 @@
+import interpreters
+import os
+from textwrap import dedent
+import unittest
+
+def _captured_script(script):
+    r, w = os.pipe()
+    indented = script.replace('\n', '\n                ')
+    wrapped = dedent(f"""
+        import contextlib
+        with open({w}, 'w') as spipe:
+            with contextlib.redirect_stdout(spipe):
+                {indented}
+        """)
+    return wrapped, open(r)
+
+def clean_up_interpreters():
+    for id in interpreters.list_all():
+        if id == 0:  # main
+            continue
+        try:
+            interpreters.destroy(id)
+        except RuntimeError:
+            pass  # already destroyed
+
+
```

Comment:
For unit tests (i.e. not "functional") I recommend using a separate test double for the low-level module:

```suggestion

class LowLevelStub:
    # set these as appropriate in tests
    errors = ()
    return_create = ()
    ...
    def __init__(self):
        self._calls = []
    def _add_call(self, name, args=(), kwargs=None):
        self.calls.append(
            (name, args, kwargs or {}))
    def _maybe_error(self):
        if not self.errors:
            return
        err = self.errors.pop(0)
        if err is not None:
            raise err
    def check_calls(self, test, expected):
        test.assertEqual(self._calls, expected)
        for returns in [self.errors, self.return_create, ...]:
            test.assertEqual(tuple(returns), ())  # make sure all were used
        
    # the stubbed methods
    def create(self):
        self._add_call('create')
        self._maybe_error()
        return self.return_create.pop(0)
    def list_all(self):
        ...
    def get_current(self):
        ...
    def get_main(self):
        ...
    def destroy(self, id):
        ...
    def run_string(self, id, text, ...):
        ...


```

Then in unit tests you would make sure the code in the `interpreters` module uses the testing double instead of the actual `_interpreters` module.  This can be done by monkey-patching `interpreters._interpreters` or by having "private" keyword arguments on all the relevant functions (my preference).

FWIW, this could be done with unittest.mock, but I'm not a huge fan of automatic mocks. :)